paml/data.py

```diff
@@ -0,0 +1,66 @@
+"""
+Functions related to data i/o in connection with a protocol execution trace.
+"""
+
+import pandas as pd
+import paml_convert
+import re
+from paml_convert.autoprotocol.plate_coordinates import coordinate_rect_to_row_col_pairs, num2col
+from paml import SampleMask, SampleData, SampleArray
+from io import StringIO
+
+import logging
+l = logging.getLogger(__file__)
+l.setLevel(logging.ERROR)
+
+def sample_array_to_dataframe(self):
+    container_type = self.container_type.lookup()
+    contents = self.contents
+
+    # FIXME need to find a definition of the container topology from the type
+
+    # FIXME this assumes a 96 well plate
+    l.warn("Warning: Assuming that the SampleArray is a 96 well microplate!")
+
+    row_col_pairs = coordinate_rect_to_row_col_pairs("A1:H12")
+    aliquots = [f"{num2col(c+1)}{r}" for (r, c) in row_col_pairs]
+    sources = [self.name for i in range(len(aliquots))]
+    df = pd.DataFrame({
```

Comment:
Since everything is going to/from dataframes, and since dataframes have a nice compact serialization form, should we just adopt dataframe serialization directly as are means of representing normal arrays? The only reason to insist on reifying the arrays into RDF is if we want to be able to perform RDF queries on them, which I think would be generally not preferred.

Reply:
The solutions I found for representing CSVs or data frames were pretty clunky wrt. describing the columns using RDF.  These functions here are more of helpers to pull data into notebooks for analysis.  I think we shouldn't use this as a serialization format, unless we want to build a way to describe the tables in terms of RDF.

Reply:
I was thinking of something much simpler, actually, of just embedding a dataframe serialization as a string-valued property.

Reply:
Yes, I think that could work.  However, are data frame serializations portable?  I know that csv should work ...

Reply:
I have it working with pandas going to and from a string csv.